core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,371 @@
+                                new IllegalArgumentException("Invalid address: " + hostname, e));
+                        return;
+                    }
+                }
+            } finally {
+                records.forEach(ReferenceCountUtil::safeRelease);
+            }
+
+            if (inetAddress == null) {
+                result.completeExceptionally(
+                        new UnknownHostException("failed to receive DNS records for " + hostname));
+                return;
+            }
+            final CacheEntry entry = new CacheEntry(inetAddress, ttlMillis, questions);
+            result.complete(entry);
+            entry.scheduleCacheUpdate(ttlMillis);
+        });
+    }
+
+    @Override
+    protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise)
+            throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Closes all the resources allocated and used by this resolver.
+     *
+     * <p>Please note that this method does not clear the {@link CacheEntry} because the {@link #cache} is
+     * created by {@link RefreshingAddressResolverGroup} and shared across from all
+     * {@link RefreshingAddressResolver}s. {@link CacheEntry} is cleared when
+     * {@link RefreshingAddressResolverGroup#close()} is called.
+     */
+    @Override
+    public void close() {
+        resolverClosed = true;
+        resolver.close();
+    }
+
+    final class CacheEntry implements Runnable {
+
+        private final InetAddress address;
+        private final long ttlMillis;
+        private final List<DnsQuestion> questions;
+
+        /**
+         * No need to be volatile because updated only by the {@link #executor()}.
+         */
+        private int numAttemptsSoFar = 1;
+
+        private volatile int cacheHits;
+
+        @Nullable
+        volatile ScheduledFuture<?> cacheUpdatingScheduledFuture;
+
+        CacheEntry(InetAddress address, long ttlMillis, List<DnsQuestion> questions) {
+            this.address = address;
+            this.ttlMillis = ttlMillis;
+            this.questions = questions;
+        }
+
+        void incrementCacheHit() {
+            cacheHits++;
+        }
+
+        InetAddress address() {
+            return address;
+        }
+
+        long ttlMillis() {
+            return ttlMillis;
+        }
+
+        void scheduleCacheUpdate(long nextDelayMillis) {
+            for (;;) {
+                if (resolverClosed) {
+                    return;
+                }
+                final ScheduledFuture<?> oldFuture = futureUpdater.get(this);
+                if (oldFuture == closed) {
+                    return;
+                }
+
+                final ScheduledFuture<?> newFuture =
+                        executor().schedule(this, nextDelayMillis, TimeUnit.MILLISECONDS);
+                if (futureUpdater.compareAndSet(this, oldFuture, newFuture)) {
```

Comment:
I think `scheduleCacheUpdate` is also only possible from the `executor()` same as `attemptsSoFar` since it should be called from the same thread as `doResolve` - do we need the updater?

Reply:
`scheduleCacheUpdate` can be set by another thread when `clear()` is called.
So I think we need the updater but do not need the `for` loop. I was a little lazy to remove that. 😜 Let me fix it.